module/x/gravity/migrations/v2/store.go

```diff
@@ -0,0 +1,178 @@
+package v2
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	ethcmn "github.com/ethereum/go-ethereum/common"
+)
+
+// MigrateStore performs in-place store migrations from v1 to v2. The migration
+// includes:
+//
+// - Change all Cosmos orginated ERC20 mappings from (HEX) string to bytes.
+// - Change all validator Ethereum keys from (HEX) string to bytes.
+func MigrateStore(ctx sdk.Context, storeKey storetypes.StoreKey, cdc codec.BinaryCodec) error {
+	store := ctx.KVStore(storeKey)
+
+	// Denoms
+	if err := migrateCosmosOriginatedDenomToERC20(store); err != nil {
+		return err
+	}
+
+	if err := migrateCosmosOriginatedERC20ToDenom(store); err != nil {
+		return err
+	}
+
+	// Validators' Ethereum addresses
+	if err := migrateEthAddressByValidator(store); err != nil {
+		return err
+	}
+
+	if err := migrateValidatorByEthAddressKey(store); err != nil {
+		return err
+	}
+
+	// Batch confirmations
+	if err := migrateBatchConfirms(store); err != nil {
+		return err
+	}
+
+	if err := migrateOutgoingTxs(store); err != nil {
+		return err
+	}
+
+	if err := migrateOutgoingTxBatches(store); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func migrateCosmosOriginatedDenomToERC20(store storetypes.KVStore) error {
+	prefixStore := prefix.NewStore(store, []byte(DenomToERC20Key))
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		addrStr := string(iterator.Value())
+		addr := ethcmn.HexToAddress(addrStr)
+
+		prefixStore.Set(iterator.Key(), addr.Bytes())
+	}
+
+	return nil
+}
+
+func migrateCosmosOriginatedERC20ToDenom(store storetypes.KVStore) error {
+	prefixStore := prefix.NewStore(store, []byte(ERC20ToDenomKey))
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		oldKey := iterator.Key()
+
+		newKey := string(ethcmn.HexToAddress(string(oldKey)).Bytes())
+
+		prefixStore.Delete(oldKey)
+		prefixStore.Set([]byte(newKey), iterator.Value())
+	}
+
+	return nil
+}
+
+func migrateEthAddressByValidator(store storetypes.KVStore) error {
+	prefixStore := prefix.NewStore(store, []byte(EthAddressByValidatorKey))
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		addrStr := string(iterator.Value())
+		addr := ethcmn.HexToAddress(addrStr)
+		prefixStore.Set(iterator.Key(), addr.Bytes())
+	}
+
+	return nil
+}
+
+func migrateValidatorByEthAddressKey(store storetypes.KVStore) error {
+	// TODO: There's a chance that we have duplicated keys here. We should
+	// keep only the well encoded key and discard the other. This is possible
+	// given that the all lower case keys were never accepted and no confirms nor
```

Comment:
how would a duplicated key get into this code? Not opposed to a duplicates check just wondering about the thought process behind this todo.

Reply:
In `SetOrchestratorAddress` we check for duplicates using a simple `==` between 2 strings.

```
	// check that neither key is a duplicate
	delegateKeys := k.GetDelegateKeys(ctx)
	for i := range delegateKeys {
		if delegateKeys[i].EthAddress == addr.GetAddress() {
			return nil, sdkerrors.Wrap(err, "Duplicate Ethereum Key")
		}
```

As these were previously simple strings, it means that someone could have registered:
```
0x00192Fb10dF37c9FB26829eb2CC623cd1BF599E8
0x00192fb10df37c9fb26829eb2cc623cd1bf599e8
0x00192Fb10dF37c9FB26829eb2CC623cd1BF599E8
...
```

Given that any non checksummed address was being rejected while checking the signature (in `confirmHandlerCommon` https://github.com/Gravity-Bridge/Gravity-Bridge/blob/5e630adb107284d57b955d4a20e42d10167d6607/module/x/gravity/types/ethereum_signer.go#L69), we can safely delete them.

Other notes I took while replying this and thinking about it:

- I think the comment `nor claims were made.` is wrong as claims might have been made (we don't check the Eth address there).
- We might also need to remove any duplicate `EthAddressByValidatorKey`

Reply:
The migrations should be re-reviewed with all this in mind, we can add detections for edge cases and just not handled them yet, chances are many might not exist ,but we def want to know if they do.